Add HSM Segwit support RSKIP (#428)
rsksmart/RSKIPs: IPs/RSKIP428.md

---
rskip: 428
title: Emit a new event with the utxo outpoints when a peg-out is created
created: 23-APR-24
author: NC
purpose: Sca
layer: Core 
complexity: 1
status: Draft
description: 
---

|RSKIP          | 428                                                                |
| :------------ |:-------------------------------------------------------------------|
|**Title**      | Emit a new event with the utxo outpoints when a peg-out is created |
|**Created**    | 23-APR-24                                                          |
|**Author**     | NC                                                                 |
|**Purpose**    | Sca                                                                |
|**Layer**      | Core                                                               |
|**Complexity** | 1                                                                  |
|**Status**     | Draft                                                              |

## Abstract

> This RSKIP proposes creating a new event that stores the outpoint values of the inputs used in a peg-out to enable the PowHSM sign segwit peg-outs.

Reviewer: ```suggestion
This RSKIP proposes creating a new event to be emitted when a peg-out transaction is created. This event will include the outpoint values of the inputs used in the peg-out transaction as well as the hash of the Bitcoin transaction without signatures. This information will then be used by the PowHSM to build the sighash and sign segwit pegout transactions.
```

## Motivation

> For the PowHSM sign segwit peg-outs[1], the outpoint values of the inputs must be provided since they are needed for calculating the sighash of a segwit bitcoin transaction. Currently, selected UTXOs for a peg-out are removed from the available utxos set once the peg-out is created, meaning they cannot be retrieved later. Storing the outpoint values in an event, when peg-outs are created and before the selected utxos are removed from the set, is a solution that will allow the PowPeg to retrieve and provide the required outpoint values to the PowHSM to sign segwit peg-outs.

Reviewer: ```suggestion
For the PowHSM to sign segwit peg-outs[1], the outpoint values of the inputs must be provided since they are needed for calculating the sighash of a segwit bitcoin transaction. Currently, selected UTXOs for a peg-out are removed from the available utxos set in the Bridge once the peg-out is created, meaning they cannot be retrieved later. Storing the outpoint values in an event, when peg-outs are created and before the selected utxos are removed from the set, is a solution that will allow the PowPeg to retrieve and provide the required outpoint values to the PowHSM to sign segwit peg-outs.
```

## Specification

> The proposal is to create a new event to store the bitcoin transaction hash and the UTXO outpoint values of a transaction spending funds from the federation, which from now on we will call it peg-out. This event will be emitted when a peg-out is created.

Reviewer: ```suggestion
The proposal is to create a new event to store the bitcoin transaction hash and the UTXO outpoint values of a transaction spending funds from the PowPeg address. This could be either a peg-out, a migration, or a refund transaction. The event will be emitted whenever any of these transactions are created.
```

> This new event is composed of two parameters:

Reviewer: I'd put the event signature here

> TThe proposal is to create a new event to store the bitcoin transaction hash of the peg-out and the outpoint values of each utxo used as input of the given peg-out. Then, this event will be emitted whenever a transaction spending from the federation address is created. In other words, when a peg-out is created.

Reviewer 2: typo in `TThe`

Author: Fixed.

> TThe proposal is to create a new event to store the bitcoin transaction hash of the peg-out and the outpoint values of each utxo used as input of the given peg-out. Then, this event will be emitted whenever a transaction spending from the federation address is created. In other words, when a peg-out is created.

Reviewer 2: `Then, this event will be emitted whenever a transaction spending from the federation address is created. In other words, when a peg-out is created.`

i would stay with the _transaction spending from the federation address is created_ part.

Author: Updated the wording of the paragraph.

Reviewer 2: We may should rewrite the whole paragraph, to just talk about txs being spent from the powpeg.
Because the way it is now, we talk about pegouts first

Author: The last sentence states that I'm referring to a peg-out. I expressed it that way just to avoid sounding strange by mentioning each type of peg-out transaction, as Marcos pointed out above.

> First, the bitcoin transaction hash, must be the hash of the peg-out without witnesses and signatures.

Reviewer: ```suggestion
First, the bitcoin transaction hash, must be the hash of the peg-out without witnesses and without signatures.
```
Just ot be clear

> Second, the list of utxo outpoint values of the peg-out. This will be a list of one or more integer values in VarInt format[2], ordered in the same order as the inputs of the peg-out.

Reviewer: ```suggestion
Second, the list of UTXO outpoint values of the spend transaction. This will be a list of one or more integer values in VarInt format[2], ordered in the same order as the inputs of the peg-out.
```
Are the values in satoshis? Btc?

Author: Satoshis

> Second, the list of utxo outpoint values. When dealing with non-segwit peg-outs, the outpoint values will be an empty list. When dealing with segwit peg-outs, this will be a list of one or more integer  values in VarInt format[2], ordered in the same order as the inputs of the peg-out.

Reviewer 2: typo here (two spaces between `integer values`)

Author: Updated!

> Therefore, when the PowPeg node processes the peg-outs to sign, it will follow a straightforward process. It will retrieve the outpoint values from the new event in the rsk transaction where the peg-out was created. Then these values will be provided to the PowHSM to sign the given segwit peg-out.

Reviewer: I'd probably remove this paragraph. I think it exceeds the proposal of the RSKIP

### Implementation

`pegout_transaction_created` is the name of the new event that will be created. Its signature is:

```
pegout_transaction_created(bytes32 indexed btcTxHash, bytes utxoOutpointValues)
```

> As we see, this event consists of two params:

Reviewer: This section basically repeats the same as above

> `segwit_pegout_created` is the name of the new event that will be created. Its signature is:

Reviewer 2: i have a general question / concern.
If the event is called `segwit_pegout_created`, why should we deal with non-segwit peg-outs?

For me, it should be like this:
1. if we are using the event just for segwit-related purposes, let's avoid dealing with non-segwit peg-outs and empty lists. This of course implies checking, before emitting the event, which type of pegout we are dealing with.
2. if we are using the event for pegouts in general, we should change the event name and provide the outpoint values in both cases

Author: Updated!

---
rskip: 428
title: Emit a new event with the utxo outpoints when a peg-out is created
created: 23-APR-24
author: NC
purpose: Sca
layer: Core 
complexity: 2
status: Draft
description: 
---

|RSKIP          | 428                |
| :------------ |:-------------------|
|**Title**      | Emit a new event with the utxo outpoints when a peg-out is created |
|**Created**    | 23-APR-24          |
|**Author**     | NC                 |
|**Purpose**    | Sca                |
|**Layer**      | Core               |
|**Complexity** | 2                  |
|**Status**     | Draft              |

## Abstract

This RSKIP proposes creating a new event that stores the outpoint values of the inputs used in a 
peg-out to enable the PowHSM sign segwit peg-outs.

> ## Motivation

Reviewer 3: Not sure, but it might be worth it to include a brief explanation on why the outpoint values are needed for the PowHSM to sign segwit pegouts

Author: Done

> In order to the PowHSM sign segwit peg-outs[1], it needs the outpoint values to be provided when 

Reviewer 3: ```suggestion
In order for the PowHSM to sign segwit peg-outs[1], it needs the outpoint values to be provided when 
```

Author: Update

---
rskip: 428
title: HSM Segwit Support
created: 23-APR-24
author: NC
purpose: Sca
layer: Core 
complexity: 2
status: Draft
description: 
---

|RSKIP          | 428                |
| :------------ |:-------------------|

> |**Title**      | Support HSM Segwit |

Author: What about "_Emit a new event with the utxo outpoints when a peg-out is created_"?

|**Created**    | 23-APR-24          |
|**Author**     | NC                 |
|**Purpose**    | Sca                |
|**Layer**      | Core               |

> |**Complexity** | 2                  |

Author: Yes, create a new event and use it in the places where it should be emitted.

> |**Complexity** | 2                  |

Reviewer: Doesn't sound too complex then :)

|**Status**     | Draft              |

## Abstract

This RSKIP proposes creating a new event that stores the outpoint values of the inputs used in a 

> peg-out to enable HSM Segwit support in RSK.

Reviewer: `HSM Segwit support` sounds a bit ambiguous. I'd try to be more clear. Explain that the goal is to provide the utxos value so that the PowHSM can build the sighash of a segwit transaction in order to sign it

## Motivation

To support HSM Segwit, the PowPeg must provide the outpoint values for signing a segwit peg-out[1].
Currently, selected UTXOs for a peg-out are removed from the available utxos set once 
the peg-out is created. Storing the outpoint values in an event will allow the PowPeg to retrieve 
and provide the outpoint values to the HSM, so being able to sign segwit peg-out properly.

## Specification

The proposal is to create a new event to store the Bitcoin Transaction Hash of the peg-out and 

> the outpoint values of each utxo used for it. This event should be emitted whenever a peg-out is 

Reviewer: Maybe instead of `peg-out` call it `a transaction that spends from the PowPeg address`? Or something similar. Sounds a bit strange to say that a peg-ut could be a migration or a refund transaction

created, including user peg-outs, migrations, and refunds for rejected transactions.

As described, this new event is composed of two parameters:

> First, the Bitcoin Transaction Hash, must be the hash of the peg-out without witnesses and 

Reviewer: `Bitcoin Transaction Hash` why the camel case?

> whenever a transaction that spends from the federation address, in other words peg-outs, is created.

Reviewer 2: ` in other words **a peg-out**` (singular, not plural, since you are talking about **a** transaction)

signatures.

Second, the list of outpoint values. When dealing with non-segwit peg-outs, the outpoint values will
be an empty list. When dealing with segwit peg-outs, this will be a list of one or more integer 
values in VarInt format[2], ordered in the same order as the inputs of the peg-out transaction.

So, when the PowPeg processes the peg-outs to sign, it will follow a straightforward process. 
It will retrieve the outpoint values from the new event in the rsk transaction where the peg-out 
was created. These values are then provided to the HSM, facilitating the signing of the segwit 
peg-outs.

### Implementation

> `SEGWIT_PEGOUT_CREATED` is the name of the new event that will be created. Its signature is:

Reviewer 4: ```suggestion
`segwit_pegout_created ` is the name of the new event that will be created. Its signature is:
```

```
segwit_pegout_created(bytes32 indexed btcTxHash, bytes utxoOutpointValues)
```

As we see, this event consists of two params:

- `btcTxHash` parameter represents the transaction hash without signatures of the pegout or 
migration transaction created. This param is limited to 32 bytes, which is the size of any Bitcoin
transaction hash. This parameter is indexed and allows search queries by a given Bitcoin transaction hash.

- `utxoOutpointValues` parameter represents the outpoint values used for the pegout transaction. 
The order must match the inputs order for the pegout transaction. This parameter stores a list of 
integers in VarInt format serialized as a byte array. The list of outpoint values is serialized and
deserialized using VarInt encoding/decoding mechanism. The order of the outpoint values is preserved
when serializing and deserializing the list of values.

> This event should be emitted every time a new peg-out is created.

Reviewer: same here

## Rationale

> To make PowPeg able to retrieve the outpoint values used in a segwit peg-out transaction, so them can be 

Reviewer 4: ```suggestion
To make the PowPeg nodes able to retrieve the outpoint values used in a segwit peg-out transaction, so they can be 
```
